### PR TITLE
feat - GPTSFTChatDataset alignment with OpenAI Messages, compatibility with packed sequences

### PR DESCRIPTION
**Pull request by soluwalana**
> [!IMPORTANT]  
> The `Update branch` button must only be pressed in very rare occassions.
> An outdated branch is never blocking the merge of a PR.
> Please reach out to the automation team before pressing that button.

# What does this PR do ?

Updates the GPTSFTChatDataset to use the messages format from OpenAI
Updates sequence packing to be compatible with GPTSFTChatDataset

**Collection**: llm

# GitHub Actions CI

The Jenkins CI system has been replaced by GitHub Actions self-hosted runners.

The GitHub Actions CI will run automatically when the "Run CICD" label is added to the PR.
To re-run CI remove and add the label again.
To run CI on an untrusted fork, a NeMo user with write access must first click "Approve and run".

# Before your PR is "Ready for review"

**Pre checks**:

- [ ] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:

- [ ] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed.
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information

- Related to # (issue)

### CODE REVIEW

File: nemo/collections/common/tokenizers/huggingface/auto_tokenizer.py [hunk 1]
```diff
@@ -46,6 +46,7 @@ def __init__(
         use_fast: Optional[bool] = False,
         trust_remote_code: Optional[bool] = False,
         include_special_tokens: bool = False,
+        chat_template: Optional[str] = None,
```
Comment (athitten) [inline nemo/collections/common/tokenizers/huggingface/auto_tokenizer.py:49]: Nit: Can we add chat_template arg in the docstrings ?

Comment (soluwalana) [reply nemo/collections/common/tokenizers/huggingface/auto_tokenizer.py:49]: Of course!

File: nemo/collections/common/tokenizers/sentencepiece_tokenizer.py [hunk 2]
```diff
@@ -221,8 +281,8 @@ def ids_to_text(self, ids):
 
             for i, id in enumerate(ids):
                 if id in self.id_to_special_token:
-                    text += self.tokenizer.decode_ids(ids[last_i:i]) + " "
-                    text += self.id_to_special_token[id] + " "
+                    text += self.tokenizer.decode_ids(ids[last_i:i])
+                    text += self.id_to_special_token[id]
                     last_i = i + 1
```
Comment (athitten) [inline nemo/collections/common/tokenizers/sentencepiece_tokenizer.py:286]: Unsure why we had `+ " "` initially, don't we need it anymore ?

File: nemo/collections/llm/gpt/data/core.py [hunk 3]
```diff
@@ -856,89 +873,179 @@ def collate_fn(self, batch):
...
         """
-        result = _preprocess(
-            example,
-            self.tokenizer,
-            self.name_end_token_ids,
-            self.label_start_tokens,
-            self.special_tokens,
-            self.num_turn_start_tokens,
-        )
+        result = _chat_preprocess(example, self.tokenizer, self.tool_schemas)
```
Comment (athitten) [inline nemo/collections/llm/gpt/data/core.py:1005]: Is `_chat_preprocess` also taking care of the preprocessing that happened in `_preprocess` ?

File: nemo/export/onnx_llm_exporter.py [hunk 4]
```diff
@@ -55,6 +57,14 @@ def wrapper(*args, **kwargs):
...
 
 
+use_trt = True
+try:
+    import tensorrt as trt
+except ImportError:
+    logging.warning("tensorrt is not available")
+    use_trt = False
+
+
```
Comment (athitten) [inline nemo/export/onnx_llm_exporter.py:67]: Why is this added ? Where is it getting used ?

**Comment by athitten** [comment]
I reviewed that parts that I could and left comments, but I feel for changes in `GPTSFTChatDataset` class, `nemo/collections/llm/gpt/data/packed_sequence.py`, `nemo/utils/sequence_packing_utils.py` its good to get review from @cuichenx who is the PIC for SFT and packed sequence in NeMo. FYI Chen is OOO and back on Wednesday.

Also for `nemo/collections/llm/gpt/model/llama_nemotron.py` @suiyoubi could you please take a look ? Thanks!

File: nemo/collections/llm/gpt/data/core.py (hunk 3 above)
Comment (soluwalana) [reply nemo/collections/llm/gpt/data/core.py:1005]: Yes, it was the original preprocess function but is relying on the apply_chat_template function for most of the functionality, whereas the original calculated many things manually, and was very error prone.

File: nemo/export/onnx_llm_exporter.py (hunk 4 above)
Comment (soluwalana) [reply nemo/export/onnx_llm_exporter.py:67]: `use_trt` variable can be removed, it was causing issues with exporting when trt isn't installed. Will update!

Comment (soluwalana) [reply nemo/export/onnx_llm_exporter.py:67]: It was just following the same pattern as pytriton and onnxruntime above it. should I also remove the unused pytriton variable?

File: nemo/utils/sequence_packing_utils.py [hunk 5]
```diff
@@ -212,21 +220,31 @@ def fill_packing_strategy(
...
-                        [
-                            # (x['answer_start_idx'] - 1) because we want to train on the output
-                            # after the last context token
-                            idx >= (x['answer_start_idx'] - 1) and x['input_ids'][idx] != pad_id
-                            for idx in range(len(x['input_ids']))
-                        ]
-                        for x in per_seq_data
-                    ]
-                )[perm].tolist()
+                loss_mask = np.array([x["mask"] for x in per_seq_data])[perm].tolist()
```
Comment (cuichenx) [inline nemo/utils/sequence_packing_utils.py:225]: Just to make sure, is the "mask" key only for chat dataset? If so could you add a comment here to explain

Comment (soluwalana) [reply nemo/utils/sequence_packing_utils.py:225]: Good call out, will ensure it is loss_mask for all modalities, updating now.

**Review by cuichenx — APPROVED** [review]
Thanks, all LGTM

File: nemo/export/onnx_llm_exporter.py (hunk 4 above)
Comment (athitten) [reply nemo/export/onnx_llm_exporter.py:67]: Thanks @soluwalana, we can leave it as is. Looks pretty harmless.

**Review by athitten — APPROVED** [review]
Thanks for addressing all comments @soluwalana ! LGTM

**Comment by jenchen13** [comment]
just merged #13273. when you rebase you can make `use_hf_tokenizer_chat_template: bool = True` in ChatDataModule to enable chat template by default for tool calling. Would also be good to update the ChatDataModule documentation to specify the new default behavior. 

Also I think you also return `context_ids` and `answer_ids` for each sample, but these fields are not used during SFT. If you'd still like to add them they can be added in [`utils.py`](https://github.com/NVIDIA/NeMo/blob/4430b1c355ad88e2713aaedae8136e8389f6a9f0/nemo/collections/llm/gpt/data/utils.py#L855)

**Comment by soluwalana** [comment]
@jenchen13 rebase completed, functionality is merged. 
Only changed the default value in GPTSFTChatDataset and not in ChatDatamodule so that the behavior from the outer interface didn't change.

**Comment by soluwalana** [comment]
For reference the ChatDataModule that we use in Customizer is [here](https://gitlab-master.nvidia.com/aire/microservices/nmp/-/blob/main/services/customizer/src/training/nemo/datamodules/chat_enabled_datamodule.py?ref_type=heads), we base off of FinetuningDatamodule and make it compatible with sequence packing.

**Review by cuichenx — DISMISSED** [review]
LGTM, thanks for the rebase.

File: nemo/collections/llm/gpt/data/utils.py [hunk 6]
```diff
@@ -799,60 +799,93 @@ def _preprocess(
     context_ids = input_ids[:last_ignore_index_pos]
     answer_ids = input_ids[last_ignore_index_pos:]
 
-    return dict(input_ids=input_ids, mask=mask, context_ids=context_ids, answer_ids=answer_ids)
+    return dict(input_ids=input_ids, loss_mask=mask, context_ids=context_ids, answer_ids=answer_ids)
 
 
-def _transform_to_chat_message(source: dict[str, list]):
```
Comment (jenchen13) [inline nemo/collections/llm/gpt/data/utils.py:805]: please keep this function `_transform_to_chat_message` instead of overwriting it. you can use it as a helper function

File: nemo/collections/llm/gpt/data/core.py [hunk 7]
```diff
@@ -1003,33 +1042,55 @@ def _process_example(self, example):
...
+        contexts = [item["context_ids"] for item in batch]
+        # Extracts the assistant's response portion of the sequence, which
+        # represents the part the model is trained to generate. This helps
+        # distinguish between the input prompt and the expected model output.
+        answers = [item["answer_ids"] for item in batch]
+        # Removes the first element from the mask to align with the shifted labels,
+        # ensuring that loss is only computed for valid, predictable tokens. This
+        # prevents the model from incurring loss on tokens that were never meant to
+        # be predicted, such as user-provided context or padding.
+        loss_mask = [item["loss_mask"][1:] for item in batch]
```
Comment (jenchen13) [inline nemo/collections/llm/gpt/data/core.py:1082]: can the comments be more concise?

File: nemo/collections/llm/gpt/data/utils.py [hunk 8]
```diff
@@ -799,60 +799,93 @@ def _preprocess(
...
-    tokens = tokenizer.tokenizer.apply_chat_template(
-        hf_chat_dict['messages'],
+
+    tokenized_chat = tokenizer.tokenizer.apply_chat_template(
+        chat,
+        tools=tools,
+        tokenize=True,
         return_dict=True,
         return_assistant_tokens_mask=template_has_generation_kwd,
-        return_tensors='pt',
```
Comment (jenchen13) [inline nemo/collections/llm/gpt/data/utils.py:848]: I think this line is needed for `input_ids` to be a torch Tensor

File: nemo/collections/llm/gpt/data/utils.py [hunk 9]
```diff
@@ -799,60 +799,93 @@ def _preprocess(
...
+    # which indicates end of context and beginning of answer
+    input_ids = tokenized_chat.get("input_ids")
     if template_has_generation_kwd:
-        mask = torch.tensor(tokens['assistant_masks']).to(bool)
+        mask = torch.tensor(tokenized_chat['assistant_masks']).to(bool)
     else:
         mask = torch.ones_like(input_ids)
-    return dict(input_ids=input_ids, mask=mask)
+    if tokenizer.eos_id:
+        input_ids += [tokenizer.eos_id]
```
Comment (jenchen13) [inline nemo/collections/llm/gpt/data/utils.py:875]: does HF tokenizer not automatically add EOS?

**Review by jenchen13 — COMMENTED** [review]
thanks for rebasing!

File: tests/collections/llm/gpt/data/test_chat_dataset.py [hunk 10]
```diff
@@ -24,12 +24,102 @@
 from nemo.collections.llm.gpt.data.chat import ChatDataModule
 
 
+def mock_llama_apply_chat_template(
```
Comment (jenchen13) [inline tests/collections/llm/gpt/data/test_chat_dataset.py:27]: can you add a test that checks that the final tokens after `apply_chat_template` are correct? otherwise I don't see the purpose of adding in a mock `apply_chat_template`.

 instead of a mock function we could also add a HF tokenizer for Llama 3.2 1B in the test files and load that

File: nemo/collections/llm/gpt/data/utils.py (hunk 6 above)
Comment (soluwalana) [reply nemo/collections/llm/gpt/data/utils.py:805]: Is this function used anywhere else?

File: nemo/collections/llm/gpt/data/core.py (hunk 7 above)
Comment (soluwalana) [reply nemo/collections/llm/gpt/data/core.py:1082]: The comments are helpful for someone modifying these parameters in the future. Why do they need to be more concise?

File: nemo/collections/llm/gpt/data/utils.py (hunk 9 above)
Comment (soluwalana) [reply nemo/collections/llm/gpt/data/utils.py:875]: Not unless you explicitly add it in the template.

File: nemo/collections/nlp/modules/common/tokenizer_utils.py [hunk 11]
```diff
@@ -150,7 +163,7 @@ def get_tokenizer(
 
 
 def get_nmt_tokenizer(
-    library: str = 'sentencepiece',
```
[the remark this answers is not in the document]
Comment (soluwalana) [reply nemo/collections/nlp/modules/common/tokenizer_utils.py:153]: 👍

File: nemo/collections/llm/gpt/data/utils.py (hunk 6 above)
Comment (soluwalana) [reply nemo/collections/llm/gpt/data/utils.py:805]: I'd prefer to not have this extra function since it is not called from anywhere outside of this file.

Comment (soluwalana) [reply nemo/collections/llm/gpt/data/utils.py:805]: Also this function is missing the important injection of the system message at the front of messages.

File: nemo/collections/llm/gpt/data/utils.py (hunk 8 above)
Comment (soluwalana) [reply nemo/collections/llm/gpt/data/utils.py:848]: 👍

File: nemo/collections/llm/gpt/data/utils.py (hunk 9 above)
Comment (soluwalana) [reply nemo/collections/llm/gpt/data/utils.py:875]: Brings up a good point though, should be checking input_ids[-1] to verify that it isn't already eos_id

File: tests/collections/llm/gpt/data/test_chat_dataset.py (hunk 10 above)
Comment (soluwalana) [reply tests/collections/llm/gpt/data/test_chat_dataset.py]: Added another test that uses the llama 3 tokenizer, validates the tokens

File: nemo/collections/llm/gpt/data/utils.py (hunk 9 above)
Comment (soluwalana) [reply nemo/collections/llm/gpt/data/utils.py:878]: Added the appropriate check here.

File: nemo/collections/llm/gpt/data/utils.py (hunk 8 above)
Comment (soluwalana) [reply nemo/collections/llm/gpt/data/utils.py:848]: The collate function correctly converts the output of this function to tensors.

Comment (soluwalana) [reply nemo/collections/llm/gpt/data/utils.py:848]: Converted to tensors at the end of the function.

File: nemo/collections/llm/gpt/data/utils.py (hunk 6 above)
Comment (jenchen13) [reply nemo/collections/llm/gpt/data/utils.py:805]: adding system message is good. I meant that separating the `_transform_to_chat_message` from the rest of the apply chat template code would make the code cleaner. Also in the future if we need to transform from ShareGPT/NeMo chat format to HF elsewhere we can reuse the helper function

File: nemo/collections/llm/gpt/data/utils.py [hunk 12]
```diff
@@ -799,60 +802,97 @@ def _preprocess(
...
+        mask += [1]
+
+    if 0 in mask:
+        # traverse the list backward for first occurrence of masked token
+        context_end_idx = len(mask) - mask[::-1].index(0)
+    else:
+        context_end_idx = len(mask)
+
+    context_ids = input_ids[:context_end_idx]
```
Comment (jenchen13) [inline nemo/collections/llm/gpt/data/utils.py:887]: what do you need the `context_ids` and `answer_ids` for in SFT? I think SFT works without them too

File: tests/collections/llm/gpt/data/test_chat_dataset.py [hunk 13]
```diff
@@ -100,3 +166,362 @@ def test_create_dataset(chat_data_module, temp_dataset_dir):
...
+    loss = collated["loss_mask"][0]
+
+    tokens_copy = copy.deepcopy(tokens)
+    tokens_copy[loss == 0] = mock_tokenizer.text_to_ids("-")[0]
+
+    hf_tokenizer = mock_tokenizer.tokenizer
+    full_string = hf_tokenizer.convert_tokens_to_string(mock_tokenizer.ids_to_tokens(tokens))
+    assistant_message_only = hf_tokenizer.convert_tokens_to_string(mock_tokenizer.ids_to_tokens(tokens_copy))
+
+    assert full_string == (
```
Comment (jenchen13) [inline tests/collections/llm/gpt/data/test_chat_dataset.py:245]: thanks for adding this test! looks great

File: nemo/collections/llm/gpt/data/utils.py (hunk 12 above)
Comment (soluwalana) [reply nemo/collections/llm/gpt/data/utils.py:887]: It maintains consistency in the interface in outputs and inputs.

File: nemo/collections/llm/gpt/data/utils.py (hunk 6 above)
Comment (soluwalana) [reply nemo/collections/llm/gpt/data/utils.py:805]: The function is less than 50 lines long at this length the added overhead of another once used private function (prefixed with a _) seems unnecessary. 

But if you are blocking on it, I'm not going to die on this hill.

Comment (soluwalana) [reply nemo/collections/llm/gpt/data/utils.py:805]: Also the return value of this new function has to be a tuple which also feels unnecessary complex (chat, tools)

Comment (soluwalana) [reply nemo/collections/llm/gpt/data/utils.py:805]: Restored the separate function.

**Review by jenchen13 — APPROVED** [review]
LGTM
